File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(vector<int> a) {
    int n = a.size();
    int cnt = 0;
    vector<int> lmn(n,2e9), rmn(n,2e9);
    lmn[0] = a[0];
    for(int i=1;i<n;i++){
        lmn[i] = min(lmn[i-1], a[i]);
    }
    rmn[n-1] = a[n-1];
    for(int i=n-2;i>=0;i--){
        rmn[i] = min(rmn[i+1], a[i]);
    }

    for(int i=0;i<n;i++){
        if((rmn[i] < a[i] && lmn[i] < a[i])){
            cnt++;
        }
    }
    return n-cnt;
}

File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.java
class Solution {
    public int solution(int[] a) {
        int n = a.length;
        
        int[] lmn = new int[n+1];
        int[] rmn = new int[n+1];
        for(int i=0;i<n;i++) lmn[i] = rmn[i] = Integer.MAX_VALUE;
        lmn[0] = a[0];
        rmn[n-1] = a[n-1];
        for(int i=1;i<n;i++) lmn[i] = Math.min(lmn[i-1], a[i]);
        for(int i=n-2;i>=0;i--) rmn[i] = Math.min(rmn[i+1], a[i]);
        
        int cnt = 0;
        for(int i=0;i<n;i++){
            if(rmn[i] < a[i] && lmn[i] < a[i]) cnt++;
        }
        return n-cnt;
        
    }
}